### PR TITLE
Implement the new bounding box API

### PR DESCRIPTION
**Pull request by LukeWood**

### CODE REVIEW

File: keras_cv/bounding_box/__init__.py [hunk 1]
```diff
@@ -12,14 +12,34 @@
...
-CONFIDENCE = 5
+
+class XYXY:
+    LEFT = 0
+    TOP = 1
+    RIGHT = 2
+    BOTTOM = 3
+    CLASS = 4
+    CONFIDENCE = 5
+
```
Comment (kartik4949) [inline keras_cv/bounding_box/__init__.py:28]: extended_corners format, 
one of the KPI needed it guess.

Comment (LukeWood) [reply keras_cv/bounding_box/__init__.py:28]: This can be done in a follow up 👍

Comment (kartik4949) [reply keras_cv/bounding_box/__init__.py:28]: yup, make sense, can be done in next pr :)

**Review by divyashreepathihalli — APPROVED** [review]

File: keras_cv/bounding_box/converters_text.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
+# Copyright 2022 The KerasCV Authors
```
Comment (quantumalaviya) [inline keras_cv/bounding_box/converters_text.py:1]: Shouldn't this file rather be named `converters_test.py` instead of `converters_text.py`?

Comment (LukeWood) [reply keras_cv/bounding_box/converters_text.py:1]: Wow good catch, thanks

**Comment by LukeWood** [comment]
gentle ping @qlzh727 @divyashreepathihalli and @fchollet - lets prioritize this PR if we can.

File: keras_cv/metrics/coco/mean_average_precision.py [hunk 3]
```diff
@@ -154,15 +159,19 @@ def update_state(self, y_true, y_pred, sample_weight=None):
         if isinstance(y_pred, tf.RaggedTensor):
             y_pred = y_pred.to_tensor(default_value=-1)
 
-        y_true = tf.cast(y_true, self.compute_dtype)
-        y_pred = tf.cast(y_pred, self.compute_dtype)
+        y_true = bounding_box.transform_format(
```
Comment (qlzh727) [inline keras_cv/metrics/coco/mean_average_precision.py:162]: Is the casting still needed here?

File: keras_cv/bounding_box/converters_text.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+# limitations under the License.
+
+import itertools
+
+import tensorflow as tf
+from absl.testing import parameterized
+
+from keras_cv import bounding_box
+
+xyxy_box = tf.constant([[10, 10, 110, 110], [20, 20, 120, 120]], dtype=tf.float32)
```
Comment (qlzh727) [inline keras_cv/bounding_box/converters_text.py:22]: Let's add a test for int dtype here.

**Review by qlzh727 — CHANGES_REQUESTED** [review]
Sorry for the late reply.

File: keras_cv/bounding_box/converters.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    "center_xywh": _corners_to_center_xywh,
+    "xyxy": _xyxy_no_op,
+}
+
+
+def transform_format(boxes, source, target, images=None):
+    """Converts bounding_boxes from one format to another.
+
+    Supported formats are:
+    - xyxy, also known as `corners` format.  In this format the first four axes
```
[the remark this answers is not in the document]
Comment (fchollet) [reply keras_cv/bounding_box/converters.py:78]: Format the names as strings with quotes and backticks.

File: keras_cv/bounding_box/converters.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        raise ValueError(
+            f"`transform_format()` received an unsupported format for the argument "
+            f"`source`.  `source` should be one of {to_corners_converters.keys()}. "
+            f"Got `source`={source}"
+        )
+    if target not in from_corners_converters:
+        raise ValueError(
+            f"`transform_format()` received an unsupported format for the argument "
+            f"`target`.  `target` should be one of {from_corners_converters.keys()}. "
+            f"Got `target`={target}"
```
Comment (fchollet) [inline keras_cv/bounding_box/converters.py:118]: Same

File: keras_cv/bounding_box/converters.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    percentages of the host image.  `images` may be a ragged Tensor.  Note that using a
+    ragged Tensor for images may cause a substantial performance loss, as each image
+    will need to be processed separately due to the mismatching image shapes.
+
+    Usage:
+    ```python
+    boxes = load_coco_dataset()
+    boxes_in_xywh = keras_cv.bounding_box.transform_format(
+        boxes,
+        original='xyxy',
```
Comment (fchollet) [inline keras_cv/bounding_box/converters.py:99]: That should be "source" not "original".

File: keras_cv/bounding_box/__init__.py [hunk 8]
```diff
@@ -12,14 +12,34 @@
...
-# Confidence exists only on y_pred, and is in the 6th index.
-CONFIDENCE = 5
+
+class XYXY:
+    LEFT = 0
+    TOP = 1
+    RIGHT = 2
+    BOTTOM = 3
+    CLASS = 4
+    CONFIDENCE = 5
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/bounding_box/__init__.py:27]: I think we have to standardize on some axes here.  To make use of our Loss functions, COCO metrics, or anything else like this we *have* to standardize on something or other.  I'd rather document them here, give users a constant to rely on, etc.

Users don't *have* to include these to use our bounding box conversion utility, as long as every axis after the `N` required to store the bounding box itself come first.  

In short, I think we need to offer these constants, but should write utilities with the expectation some users will not provide these.  Losses, metrics, layers, etc will need these though so I think we do need these.

File: keras_cv/bounding_box/converters.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+from_corners_converters = {
+    "xywh": _corners_to_xywh,
+    "center_xywh": _corners_to_center_xywh,
+    "xyxy": _xyxy_no_op,
+}
+
+
+def transform_format(boxes, source, target, images=None):
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/bounding_box/converters.py:74]: I think I like the idea of having a public namespace `bounding_box`. 

`bounding_box.transformat_format()` is a nice API name, but `transform_format()` is not.

I guess it depends on how we think users will handle their imports.  If they use the Keras style:
```
from keras_cv import bounding_box
```
or

```
from keras_cv import bounding_box as bounding_box_utils
```

then this `convert_format` or `transform` or `convert` is nice, otherwise having bounding_box in the name is a bit nicer.

Thoughts?

File: keras_cv/bounding_box/converters_text.py (hunk 4 above)
Comment (LukeWood) [reply keras_cv/bounding_box/converters_text.py]: Maybe we should assert that the inputs are a float?

File: keras_cv/bounding_box/converters.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    Supported formats are:
+    - xyxy, also known as `corners` format.  In this format the first four axes
+        represent [left, top, right, bottom] in that order.
+    - xyWH.  In this format the first four axes represent [left, top, width, height].
+    - center_xyWH.  In this format the first two coordinates represent the x and y
+        coordinates of the center of the bounding box, while the last two represent
+        the width and height of the bounding box.
+
+    Formats are case insensitive.  It is recommended that you capitalize width and
+    height to maximize the visual difference between `xyWH` and `xyxy`.
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/bounding_box/converters.py:86]: yeah, it is optional.   Users can just pass `xywh` - it gets lower cased anyways.  I suggested this because I thought it may improve readability, but we can remove it.

File: keras_cv/bounding_box/converters.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/bounding_box/converters.py]: Done

File: keras_cv/bounding_box/converters.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/bounding_box/converters.py]: We can remove this if we think it hurts readability

File: keras_cv/metrics/coco/mean_average_precision.py (hunk 3 above)
Comment (LukeWood) [reply keras_cv/metrics/coco/mean_average_precision.py:162]: Updated to do it internally, this is required for many conversions.

**Review by qlzh727 — APPROVED** [review]
LGTM. Just some minor comments.

File: keras_cv/bounding_box/converters.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+their formats.
+"""
+
+import tensorflow as tf
+
+
+# Internal exception to propagate the fact images was not passed to a converter that
+# needs it
+class RequiresImagesException(Exception):
+    pass
```
[the remark this answers is not in the document]
Comment (fchollet) [reply keras_cv/bounding_box/converters.py:31]: This will never be surfaced to users, right?

In general we should never raise a custom exception to users. But if we're sure that it will be caught internally and reraised as a regular exception this is fine.

File: keras_cv/bounding_box/converters.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (fchollet) [reply keras_cv/bounding_box/converters.py]: In Keras we used to have `keras.utils.image_utils` and so on, but we ended up moving everything to a flat `keras.utils` namespace. Overall, no strong opinion here, at least `keras_cv.bounding_box` is not nested. We can keep `keras_cv.bounding_box.convert_format`.

I'd definitely favor "convert" over "transform" though.

File: keras_cv/bounding_box/converters.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+            to specify the destination format of the `boxes` parameter.
+        images: A batch of images aligned with `boxes` on the first axis.  Should be at
+            least 3 dimensions, with the first 3 dimensions representing:
+            `[batch_size, height, width]`.  Used in some converters to compute relative
+            pixel values of the bounding box dimensions.
+        dtype: the data type to use when transforming the boxes.  Defaults to
+            `tf.float32`.
+    """
+    source = source.lower()
+    target = target.lower()
```
Comment (tanzhenyu) [inline keras_cv/bounding_box/converters.py:163]: no special case for source == target?

File: keras_cv/bounding_box/converters_test.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+# limitations under the License.
+
+import itertools
+
+import tensorflow as tf
+from absl.testing import parameterized
+
+from keras_cv import bounding_box
+
+xyxy_box = tf.constant([[10, 10, 110, 110], [20, 20, 120, 120]], dtype=tf.float32)
```
[the remark this answers is not in the document]
Comment (tanzhenyu) [reply keras_cv/bounding_box/converters_test.py:22]: +1 on ragged test

**Review by tanzhenyu — COMMENTED** [review]
Thanks for the PR!

File: keras_cv/metrics/coco/numerical_tests/recall_correctness_test.py [hunk 14]
```diff
@@ -119,8 +125,8 @@ def load_samples(fname):
     y_true = npzfile["arr_0"].astype(np.float32)
     y_pred = npzfile["arr_1"].astype(np.float32)
 
-    y_true = bounding_box.convert_to_corners(y_true, format="coco")
-    y_pred = bounding_box.convert_to_corners(y_pred, format="coco")
+    y_true = bounding_box.transform_format(y_true, source="xywh", target="xyxy")
+    y_pred = bounding_box.transform_format(y_pred, source="xywh", target="xyxy")
```
Comment (LukeWood) [inline keras_cv/metrics/coco/numerical_tests/recall_correctness_test.py:129]: favor convert terminology over transform terminology

File: keras_cv/bounding_box/converters.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/bounding_box/converters.py:31]: This will never be raised to users, it is an implementation detail only.

File: keras_cv/bounding_box/converters.py [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+    Args:
+        boxes: tf.Tensor representing bounding boxes in the format specified in the
+            `source` parameter.  `boxes` can optionally have extra dimensions stacked on
+             the final axis to store metadata.  boxes should be a 3D Tensor, with the
+             shape `[batch_size, num_boxes, *]`.
+        source: One of {" ".join([f'"{f}"' for f in TO_XYXY_CONVERTERS.keys()])}.  Used
+            to specify the original format of the `boxes` parameter.
+        target: One of {" ".join([f'"{f}"' for f in TO_XYXY_CONVERTERS.keys()])}.  Used
+            to specify the destination format of the `boxes` parameter.
+        images: A batch of images aligned with `boxes` on the first axis.  Should be at
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/bounding_box/converters.py]: Updated docstring.

File: keras_cv/bounding_box/converters.py (hunk 12 above)
Comment (LukeWood) [reply keras_cv/bounding_box/converters.py:164]: Good catch, added

File: keras_cv/bounding_box/converters_test.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/bounding_box/converters_test.py:22]: > Can we update the unit test to let images has different number of bbox within a batch (ragged)? or we have to ask user to pad the bbox for now?

Let's leave this as dense only for now to unblock the efforts blocked on this PR.

 I opened https://github.com/keras-team/keras-cv/issues/478 to track.
